File: cdp/frontend/src/api/modules/project.ts
import apiClient from "../axios";
import type { ApiResponse } from "../types";

export interface Project {
  id: number;
  name: string;
  code: string;
  description: string | null;
  status: "active" | "inactive" | "archived";
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  code: string;
  description?: string;
}

export interface ProjectUpdate {
  name?: string;
  description?: string;
  status?: "active" | "inactive" | "archived";
}

export interface ProjectMember {
  id: number;
  project_id: number;
  user_id: number;
  role: "admin" | "member";
  created_at: string;
  user?: {
    id: number;
    username: string;
    phone?: string;
  };
}

export interface ProjectMemberCreate {
  user_id: number;
  role?: "admin" | "member";
}

export interface OrgProject {
  id: number;
  org_id: number;
  project_id: number;
  created_at: string;
  organization?: {
    id: number;
    name: string;
  };
}

export interface UserProject {
  id: number;
  name: string;
  code: string;
  status: "active" | "inactive" | "archived";
  role: "admin" | "member";
  created_at: string;
}

export const projectApi = {
  // Projects
  list: async (
    params: { page?: number; page_size?: number; status?: string } = {},
  ): Promise<{
    items: Project[];
    total: number;
    page: number;
    page_size: number;
  }> => {
    const res = await apiClient.get<
      ApiResponse<{
        items: Project[];
        total: number;
        page: number;
        page_size: number;
      }>
    >("/projects", { params });
    return res.data.data!;
  },

  getById: async (id: number): Promise<Project> => {
    const res = await apiClient.get<ApiResponse<Project>>(`/projects/${id}`);
    return res.data.data!;
  },

  create: async (data: ProjectCreate): Promise<Project> => {
    const res = await apiClient.post<ApiResponse<Project>>("/projects", data);
    return res.data.data!;
  },

  update: async (id: number, data: ProjectUpdate): Promise<Project> => {
    const res = await apiClient.put<ApiResponse<Project>>(
      `/projects/${id}`,
      data,
    );
    return res.data.data!;
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/projects/${id}`);
  },

  // Project Members
  listMembers: async (
    projectId: number,
  ): Promise<{ items: ProjectMember[]; total: number }> => {
    const res = await apiClient.get<
      ApiResponse<{ items: ProjectMember[]; total: number }>
    >(`/projects/${projectId}/members`);
    return res.data.data!;
  },

  addMember: async (
    projectId: number,
    data: ProjectMemberCreate,
  ): Promise<ProjectMember> => {
    const res = await apiClient.post<ApiResponse<ProjectMember>>(
      `/projects/${projectId}/members`,
      data,
    );
    return res.data.data!;
  },

  updateMemberRole: async (
    projectId: number,
    userId: number,
    role: "admin" | "member",
  ): Promise<ProjectMember> => {
    const res = await apiClient.put<ApiResponse<ProjectMember>>(
      `/projects/${projectId}/members/${userId}`,
      { role },
    );
    return res.data.data!;
  },

  removeMember: async (projectId: number, userId: number): Promise<void> => {
    await apiClient.delete(`/projects/${projectId}/members/${userId}`);
  },

  // Organization Associations
  listOrganizations: async (
    projectId: number,
  ): Promise<{ items: OrgProject[]; total: number }> => {
    const res = await apiClient.get<
      ApiResponse<{ items: OrgProject[]; total: number }>
    >(`/projects/${projectId}/organizations`);
    return res.data.data!;
  },

  associateOrg: async (
    projectId: number,
    orgId: number,
  ): Promise<OrgProject> => {
    const res = await apiClient.post<ApiResponse<OrgProject>>(
      `/projects/${projectId}/organizations`,
      { org_id: orgId },
    );
    return res.data.data!;
  },

  disassociateOrg: async (projectId: number, orgId: number): Promise<void> => {
    await apiClient.delete(`/projects/${projectId}/organizations/${orgId}`);
  },

  // User Projects
  getMyProjects: async (): Promise<{ items: UserProject[]; total: number }> => {
    const res = await apiClient.get<
      ApiResponse<{ items: UserProject[]; total: number }>
    >("/projects/users/me/projects");
    return res.data.data!;
  },
};

File: cdp/backend/src/app/schemas/project.py
"""Project schemas."""

import enum
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ProjectStatus(str, enum.Enum):
    active = "active"
    inactive = "inactive"
    archived = "archived"


class ProjectMemberRole(str, enum.Enum):
    admin = "admin"
    member = "member"


# Project Schemas
class ProjectCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="项目名称")
    code: str = Field(..., min_length=1, max_length=100, description="项目代码（唯一）")
    description: Optional[str] = Field(None, description="项目描述")


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[ProjectStatus] = None


class ProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    code: str
    description: Optional[str]
    status: ProjectStatus
    created_at: datetime
    updated_at: datetime


class ProjectListResponse(BaseModel):
    items: list[ProjectResponse]
    total: int
    page: int
    page_size: int


# Project Member Schemas
class ProjectMemberCreate(BaseModel):
    user_id: int = Field(..., description="用户ID")
    role: ProjectMemberRole = Field(default=ProjectMemberRole.member, description="角色")


class ProjectMemberUpdate(BaseModel):
    role: ProjectMemberRole = Field(..., description="角色")


class ProjectMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    user_id: int
    role: ProjectMemberRole
    created_at: datetime
    user: Optional["UserResponse"] = None


class ProjectMemberListResponse(BaseModel):
    items: list[ProjectMemberResponse]
    total: int


# Org Project Schemas
class OrgProjectCreate(BaseModel):
    org_id: int = Field(..., description="组织ID")


class OrgProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    org_id: int
    project_id: int
    created_at: datetime
    organization: Optional["OrgUnitResponse"] = None


class OrgProjectListResponse(BaseModel):
    items: list[OrgProjectResponse]
    total: int


# User Project (for user's project list)
class UserProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    code: str
    status: ProjectStatus
    role: ProjectMemberRole
    created_at: datetime


class UserProjectListResponse(BaseModel):
    items: list[UserProjectResponse]
    total: int


# Import for type hints
from app.schemas.auth import UserResponse  # noqa: E402, F401
from app.schemas.org_unit import OrgUnitResponse  # noqa: E402, F401
